packages/ai-jsx/src/lib/asr/asr.tsx

```diff
@@ -0,0 +1,428 @@
+      }
+    }
+
+    // Copy the input data to a new Float32Array
+    const data = new Float32Array(input[0]);
+
+    // Post the data back to the main thread
+    this.port.postMessage(data);
+
+    return true;
+  }
+}
+
+registerProcessor("input-processor", InputProcessor);
+`;
+
+class EventTargetImpl implements EventTarget {
+  listeners: Object;
+  constructor() {
+    this.listeners = {};
+  }
+  addEventListener(type: string, listener) {
+    if (!this.listeners[type]) {
+      this.listeners[type] = [];
+    }
+    this.listeners[type].push(listener);
+  }
+  removeEventListener(type: string, listener) {
+    const index = this.listeners[type].indexOf(listener);
+    if (index != -1) {
+      this.listeners[type].splice(index, 1);
+    }
+  }
+  dispatchEvent(e: CustomEvent) {
+    this.listeners[e.type]?.forEach((listener) => listener(e));
+    return true;
+  }
+}
+
+export class MicManager extends EventTargetImpl {
+  private outBuf?: Float32Array[];
+  private context?: AudioContext;  
+  private streamElement?: HTMLAudioElement;
+  private stream?: MediaStream;
+  private processorNode?: AudioWorkletNode;
+  private numSilentFrames = 0;
+  async startMic(timeslice: number) {
+    this.startGraph(
+      await navigator.mediaDevices.getUserMedia({ audio: true }),
+      timeslice,
+    );
+  }
+  async startFile(url: string, timeslice: number) {
+    const response = await fetch(url);
+    const blob = await response.blob();
+    this.streamElement = new Audio();
+    this.streamElement.src = URL.createObjectURL(blob);
+    await this.streamElement.play();
+    const stream = await (this.streamElement as any).captureStream();
+    await this.startGraph(stream, timeslice);
+  }
+  private async startGraph(stream: MediaStream, timeslice: number) {
+    this.outBuf = [];
+    this.context = new window.AudioContext(); 
+    this.stream = stream;
+    const workletSrcBlob = new Blob([AUDIO_WORKLET_SRC], {
+      type: 'application/javascript',
+    });
+    const workletSrcUrl = URL.createObjectURL(workletSrcBlob);
+    await this.context!.audioWorklet.addModule(workletSrcUrl);
+    this.processorNode = new AudioWorkletNode(this.context, 'input-processor');
+    this.processorNode.port.onmessage = (event) => {
+      this.outBuf!.push(event.data);
+      const bufDuration =
+        ((128 * this.outBuf!.length) / this.sampleRate()) * 1000;
+      if (bufDuration >= timeslice) {
+        const chunkEvent = new CustomEvent('chunk', {
+          detail: this.makePcmChunk(this.outBuf!),
+        });
+        this.dispatchEvent(chunkEvent);
+        this.outBuf = [];
+      }
+    };
+    const source = this.context.createMediaStreamSource(stream);
+    source.connect(this.processorNode);
+    this.numSilentFrames = 0;
+  }
+  stop() {
+    this.processorNode!.disconnect();
+    this.stream!.getTracks().forEach((track) => track.stop());
+    if (this.streamElement) {
+      this.streamElement.pause();
+    }
+    this.context!.close();
```

Comment:
Will this cause a crash if someone calls `stop` without calling `startGraph` first?